File: examples/openai_example.py
#!/usr/bin/env python3
"""
OpenAI integration with Toko-Mo-Co.

Usage:
    export OPENAI_API_KEY="your-key"
    python openai_example.py

Proxy must be running at http://localhost:8081
"""

import os
from openai import OpenAI

PROXY_URL = os.environ.get("TOKOMOCO_URL", "http://localhost:8081")
API_KEY = os.environ.get("OPENAI_API_KEY", "")

# -- Setup (2 lines changed from normal usage) --------------------------------
# NOTE: OpenAI SDK needs /v1 in the base_url (unlike Anthropic)
client = OpenAI(
    api_key=API_KEY,
    base_url=f"{PROXY_URL}/v1",                  # <-- point to proxy + /v1
    default_headers={                             # <-- optional tracking
        "X-Session-ID": "openai-example",
        "X-Agent-ID": "example-agent",
    },
)


# -- 1. Basic request ---------------------------------------------------------
def basic_request():
    print("=== Basic Request ===")
    response = client.chat.completions.create(
        model="gpt-4o",
        messages=[{"role": "user", "content": "What is the capital of Japan?"}],
    )
    print(f"Response: {response.choices[0].message.content}\n")


# -- 2. Streaming request -----------------------------------------------------
def streaming_request():
    print("=== Streaming Request ===")
    stream = client.chat.completions.create(
        model="gpt-4o",
        messages=[{"role": "user", "content": "Count from 1 to 5."}],
        stream=True,
    )
    for chunk in stream:
        if chunk.choices[0].delta.content:
            print(chunk.choices[0].delta.content, end="", flush=True)
    print("\n")


# -- 3. Multi-turn conversation ------------------------------------------------
def multi_turn():
    print("=== Multi-turn Conversation ===")
    messages = [
        {"role": "system", "content": "You are a helpful assistant."},
        {"role": "user", "content": "My name is Alice."},
    ]
    r1 = client.chat.completions.create(model="gpt-4o", messages=messages)
    reply1 = r1.choices[0].message.content
    print(f"Assistant: {reply1}")

    messages.append({"role": "assistant", "content": reply1})
    messages.append({"role": "user", "content": "What is my name?"})

    r2 = client.chat.completions.create(model="gpt-4o", messages=messages)
    print(f"Assistant: {r2.choices[0].message.content}\n")


# -- 4. Different models (cost comparison) -------------------------------------
def model_comparison():
    print("=== Model Cost Comparison ===")
    models = ["gpt-4o-mini", "gpt-4o"]
    for model in models:
        response = client.chat.completions.create(
            model=model,
            messages=[{"role": "user", "content": "What is 2+2?"}],
        )
        print(f"{model}: {response.choices[0].message.content}")
    print("Check the dashboard to compare costs!\n")


# -- Main ----------------------------------------------------------------------
if __name__ == "__main__":
    if not API_KEY:
        print("Set OPENAI_API_KEY environment variable first.")
        raise SystemExit(1)

    print(f"Proxy: {PROXY_URL}/v1\n")
    basic_request()
    streaming_request()
    multi_turn()
    model_comparison()
    print("Done! Check the dashboard at http://localhost:8081")

File: examples/openai_example.ts
/**
 * OpenAI integration with Toko-Mo-Co (TypeScript/Node.js).
 *
 * Usage:
 *     npm install openai
 *     export OPENAI_API_KEY="your-key"
 *     npx tsx openai_example.ts
 *
 * Proxy must be running at http://localhost:8081
 */

import OpenAI from "openai";

const PROXY_URL = process.env.TOKOMOCO_URL || "http://localhost:8081";

// -- Setup (2 lines changed from normal usage) --------------------------------
// NOTE: OpenAI SDK needs /v1 in the baseURL (unlike Anthropic)
const client = new OpenAI({
  apiKey: process.env.OPENAI_API_KEY,
  baseURL: `${PROXY_URL}/v1`, // <-- point to proxy + /v1
  defaultHeaders: {
    // <-- optional tracking
    "X-Session-ID": "ts-openai-example",
    "X-Agent-ID": "ts-example-agent",
  },
});

// -- 1. Basic request ---------------------------------------------------------
async function basicRequest() {
  console.log("=== Basic Request ===");
  const response = await client.chat.completions.create({
    model: "gpt-4o",
    messages: [{ role: "user", content: "What is the capital of Japan?" }],
  });
  console.log(`Response: ${response.choices[0].message.content}\n`);
}

// -- 2. Streaming request -----------------------------------------------------
async function streamingRequest() {
  console.log("=== Streaming Request ===");
  const stream = await client.chat.completions.create({
    model: "gpt-4o",
    messages: [{ role: "user", content: "Count from 1 to 5." }],
    stream: true,
  });
  for await (const chunk of stream) {
    const content = chunk.choices[0]?.delta?.content;
    if (content) {
      process.stdout.write(content);
    }
  }
  console.log("\n");
}

// -- 3. Multi-turn conversation -----------------------------------------------
async function multiTurn() {
  console.log("=== Multi-turn Conversation ===");
  const messages: OpenAI.ChatCompletionMessageParam[] = [
    { role: "system", content: "You are a helpful assistant." },
    { role: "user", content: "My name is Alice." },
  ];

  const r1 = await client.chat.completions.create({
    model: "gpt-4o",
    messages,
  });
  const reply1 = r1.choices[0].message.content || "";
  console.log(`Assistant: ${reply1}`);

  messages.push({ role: "assistant", content: reply1 });
  messages.push({ role: "user", content: "What is my name?" });

  const r2 = await client.chat.completions.create({
    model: "gpt-4o",
    messages,
  });
  console.log(`Assistant: ${r2.choices[0].message.content}\n`);
}

// -- Main ---------------------------------------------------------------------
async function main() {
  if (!process.env.OPENAI_API_KEY) {
    console.error("Set OPENAI_API_KEY environment variable first.");
    process.exit(1);
  }

  console.log(`Proxy: ${PROXY_URL}/v1\n`);
  await basicRequest();
  await streamingRequest();
  await multiTurn();
  console.log("Done! Check the dashboard at http://localhost:8081");
}

main().catch(console.error);
